File: apps/mobile/src/types/domain.ts
/**
 * Shared domain vocabulary.
 *
 * Entity interfaces are introduced by the phase that persists them, together
 * with their Zod schemas and Firestore converters, so runtime validation and
 * types always arrive as a pair.
 *
 */

import type { Timestamp } from 'firebase/firestore';

/** Ordered low → critical. Order is meaningful: used for alert prioritisation. */
export const RISK_LEVELS = ['low', 'medium', 'high', 'critical'] as const;
export type RiskLevel = (typeof RISK_LEVELS)[number];

/** Black spot categories, matching the BlackSpot.category field. */
export const BLACK_SPOT_CATEGORIES = ['accident', 'crime', 'mixed', 'unsafe-road'] as const;
export type BlackSpotCategory = (typeof BLACK_SPOT_CATEGORIES)[number];

/** Human-readable risk labels. Risk is never communicated by colour alone. */
export const RISK_LEVEL_LABELS: Record<RiskLevel, string> = {
  low: 'Low risk',
  medium: 'Medium risk',
  high: 'High risk',
  critical: 'Critical risk',
};

/** Sort weight for prioritising overlapping alerts (higher wins). */
export const RISK_LEVEL_WEIGHT: Record<RiskLevel, number> = {
  low: 0,
  medium: 1,
  high: 2,
  critical: 3,
};

// -----------------------------------------------------------------------------
// User profile
// -----------------------------------------------------------------------------

/**
 * Access roles.
 *
 * Only ever assigned server-side. Firestore rules pin `role` to `"user"` on
 * create and forbid changing it, so a client cannot escalate itself; promotion
 * to moderator or admin happens through the Admin SDK in Phase 7.
 */
export const USER_ROLES = ['user', 'moderator', 'admin'] as const;
export type UserRole = (typeof USER_ROLES)[number];

export const THEME_PREFERENCES = ['system', 'light', 'dark'] as const;
export type DarkModePreference = (typeof THEME_PREFERENCES)[number];

/**
 * A user's profile document at `users/{id}`.
 *
 * Deliberately minimal: no location history, no device fingerprint, nothing that
 * is not needed by a documented feature. `id` matches the Firebase Auth uid.
 *
 * Timestamps are `Timestamp` on read. They are written as
 * `serverTimestamp()` sentinels, so the server clock is authoritative rather
 * than a device clock that may be wrong or deliberately altered.
 */
export interface UserProfile {
  id: string;
  name: string;
  email: string;
  phone?: string;
  role: UserRole;
  alertRadiusM: number;
  alertsEnabled: boolean;
  backgroundMonitoringEnabled: boolean;
  hapticsEnabled: boolean;
  soundEnabled: boolean;
  darkModePreference: DarkModePreference;
  createdAt: Timestamp | null;
  updatedAt: Timestamp | null;
}

/** Fields a user may change about themselves. */
export type UserProfilePreferences = Pick<
  UserProfile,
  | 'name'
  | 'phone'
  | 'alertRadiusM'
  | 'alertsEnabled'
  | 'backgroundMonitoringEnabled'
  | 'hapticsEnabled'
  | 'soundEnabled'
  | 'darkModePreference'
>;

// -----------------------------------------------------------------------------
// Black spot
// -----------------------------------------------------------------------------

/** How a black spot came to exist. Drives how much trust the UI conveys. */
export const BLACK_SPOT_SOURCES = ['manual', 'reports', 'algorithm', 'official'] as const;
export type BlackSpotSource = (typeof BLACK_SPOT_SOURCES)[number];

/**
 * A published black spot at `blackSpots/{id}`.
 *
 * Only documents with `verified === true` **and** `active === true` are ever
 * shown to users or used for alerting. Everything else — algorithm candidates,
 * deactivated spots, records still under review — stays invisible. That rule is
 * enforced in the query, in the Firestore security rules, and again on read, so
 * no single mistake can surface an unverified hazard as an official one.
 */
export interface BlackSpot {
  id: string;
  name: string;
  description?: string;
  category: BlackSpotCategory;
  latitude: number;
  longitude: number;
  /**
   * Geohash of the coordinates, written by whoever creates the document.
   *
   * Firestore has no native radius query, so nearby lookups are done as
   * range queries over this field — see blackSpotRepository.
   */
  geohash: string;
  /** Warning radius in metres. */
  radiusM: number;
  riskLevel: RiskLevel;
  /** Normalised 0–100, produced by the analytics service in Phase 10. */
  severityScore: number;
  accidentCount: number;
  crimeCount: number;
  reportCount: number;
  verified: boolean;
  active: boolean;
  source: BlackSpotSource;
  createdBy: string;
  createdAt: Timestamp | null;
  updatedAt: Timestamp | null;
}

/** A black spot paired with the user's current distance from it. */
export interface NearbyBlackSpot {
  spot: BlackSpot;
  /** Great-circle distance from the user, in metres. */
  distanceM: number;
}

// -----------------------------------------------------------------------------
// Incident report
// -----------------------------------------------------------------------------

/** What the reporter says happened. Drives moderation routing in Phase 7. */
export const INCIDENT_TYPES = ['accident', 'crime', 'pothole', 'unsafe-road', 'other'] as const;
export type IncidentType = (typeof INCIDENT_TYPES)[number];

/**
 * Reporter-assessed severity.
 *
 * Deliberately a three-point scale, not the four-point `RiskLevel` used for
 * published black spots. A member of the public is judging one event they
 * witnessed; a black spot's risk level is a moderated, aggregated judgement.
 * Keeping the vocabularies separate stops a self-reported "critical" from ever
 * reading as an official critical-risk classification.
 */
export const INCIDENT_SEVERITIES = ['low', 'medium', 'high'] as const;
export type IncidentSeverity = (typeof INCIDENT_SEVERITIES)[number];

/**
 * Moderation state of a report.
 *
 * A client may only ever cause `pending`. `approved` and `rejected` are set by a
 * moderator through the Admin SDK in Phase 7, and Firestore rules refuse any
 * client write to the field — that is what stops the reporting flow from
 * becoming a way to publish unverified warnings.
 *
 * TODO(phase-11): `draft` is part of the agreed model but nothing writes it yet;
 * local drafts for reports composed offline are Phase 11 work.
 */
export const REPORT_STATUSES = ['draft', 'pending', 'approved', 'rejected'] as const;
export type ReportStatus = (typeof REPORT_STATUSES)[number];

/**
 * A crowdsourced report at `incidentReports/{id}`.
 *
 * An approved report is evidence *towards* a black spot, never a black spot
 * itself: nothing in this app promotes a report automatically. Approval is a
 * human decision, and publishing the resulting black spot is a second, separate
 * one (Phase 7), with the clustering that proposes candidates arriving in
 * Phase 10.
 */
export interface IncidentReport {
  id: string;
  /** Firebase Auth uid of the reporter. Pinned to the caller by the rules. */
  reporterId: string;
  type: IncidentType;
  description: string;
  latitude: number;
  longitude: number;
  /** Geohash of the coordinates, so Phase 10 can cluster without a full scan. */
  geohash: string;
  severity: IncidentSeverity;
  /** When the incident happened, if the reporter said. Not the submission time. */
  occurredAt?: Timestamp | null;
  /** Firebase Storage download URLs. Empty when no photograph was attached. */
  imageUrls: string[];
  status: ReportStatus;
  /** Moderator's explanation, shown to the reporter. Set server-side only. */
  moderationNotes?: string;
  reviewedBy?: string;
  reviewedAt?: Timestamp | null;
  createdAt: Timestamp | null;
  updatedAt: Timestamp | null;
}

// -----------------------------------------------------------------------------
// Emergency contact
// -----------------------------------------------------------------------------

/**
 * Someone the user has chosen to notify with an SOS, at
 * `emergencyContacts/{id}`.
 *
 * This is the most sensitive collection in the app: it is a list of real people
 * with real phone numbers, and they never consented to being here. It is
 * therefore deliberately minimal — a name, a number, and an optional label so
 * the user can tell two "Mum"s apart — and readable only by its owner. No email,
 * no address, no notes field that would invite storing more.
 *
 * Nothing is ever sent to these contacts automatically. The phone number is
 * handed to the device's own SMS composer, which the user must then send
 * themselves; this app has no server-side messaging and never will without a
 * separate, explicit decision.
 */
export interface EmergencyContact {
  id: string;
  /** Owner's Firebase Auth uid. Pinned by the security rules. */
  userId: string;
  name: string;
  /** Stored as the user typed it, minus separators. Never normalised to E.164. */
  phone: string;
  /** Free-text label such as "Sister" or "Neighbour". Optional. */
  relationship?: string;
  /**
   * Pre-selected when the SOS screen opens.
   *
   * At most one contact should carry this. It is enforced by the repository
   * writing the change in a batch rather than by the rules, because Firestore
   * rules cannot see other documents cheaply enough to make it a hard guarantee
   * — so the SOS screen also tolerates several, or none.
   */
  isPrimary: boolean;
  createdAt: Timestamp | null;
  updatedAt: Timestamp | null;
}

/** Fields the user supplies. The rest is derived or server-set. */
export type EmergencyContactInput = Pick<
  EmergencyContact,
  'name' | 'phone' | 'relationship' | 'isPrimary'
>;

// -----------------------------------------------------------------------------
// Alert log
// -----------------------------------------------------------------------------

export const ALERT_TYPES = ['foreground', 'background', 'push'] as const;
export type AlertType = (typeof ALERT_TYPES)[number];

/**
 * A record that the user was warned, at `alertLogs/{id}`.
 *
 * Coordinates are deliberately **optional** and are not written by the app in
 * Phase 4. Storing a position with every alert would build exactly the
 * continuous location history the project promised not to keep — the user is
 * told their location is not uploaded. The black spot id already says where the
 * alert happened, to the precision anyone needs.
 */
export interface AlertLog {
  id: string;
  userId: string;
  blackSpotId: string;
  distanceM: number;
  alertType: AlertType;
  latitude?: number;
  longitude?: number;
  createdAt: Timestamp | null;
}

File: services/analytics/app/models/domain.py
"""Domain vocabulary and models for the analytics service.

The string constants here are a **mirror** of
``packages/shared-types/src/vocabulary.ts``. They are duplicated rather than
imported because there is no sane way for Python to read a TypeScript module,
and a duplicated constant that drifts is exactly the failure this project has
been avoiding since Phase 7 — a candidate written with a risk level the mobile
app renders as nothing.

The duplication is therefore **tested**: ``tests/test_vocabulary_mirror.py``
parses the TypeScript source and asserts these lists match it exactly, so drift
fails the build instead of failing silently in production.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, field_validator

# -----------------------------------------------------------------------------
# Mirrored vocabulary
# -----------------------------------------------------------------------------


class RiskLevel(StrEnum):
    """Ordered low → critical. Order is meaningful and must not be changed."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class BlackSpotCategory(StrEnum):
    ACCIDENT = "accident"
    CRIME = "crime"
    MIXED = "mixed"
    UNSAFE_ROAD = "unsafe-road"


class BlackSpotSource(StrEnum):
    MANUAL = "manual"
    REPORTS = "reports"
    ALGORITHM = "algorithm"
    OFFICIAL = "official"


class IncidentType(StrEnum):
    ACCIDENT = "accident"
    CRIME = "crime"
    POTHOLE = "pothole"
    UNSAFE_ROAD = "unsafe-road"
    OTHER = "other"


class IncidentSeverity(StrEnum):
    """Reporter-assessed, three-point. Deliberately not the four-point RiskLevel.

    A member of the public is judging one event they witnessed; a black spot's
    risk level is a moderated, aggregated judgement. Keeping the vocabularies
    separate is what stops a self-reported "high" from ever reading as an
    official classification.
    """

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class ReportStatus(StrEnum):
    DRAFT = "draft"
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


# Collection names, mirrored from COLLECTIONS in vocabulary.ts plus the two this
# phase introduces.
COLLECTION_INCIDENT_REPORTS = "incidentReports"
COLLECTION_BLACK_SPOTS = "blackSpots"
COLLECTION_BLACK_SPOT_CANDIDATES = "blackSpotCandidates"
COLLECTION_ANALYSIS_JOBS = "analysisJobs"

# Mirrored from BLACK_SPOT_RADIUS_BOUNDS_M.
BLACK_SPOT_RADIUS_MIN_M = 50
BLACK_SPOT_RADIUS_MAX_M = 5000


# -----------------------------------------------------------------------------
# Algorithm versioning
# -----------------------------------------------------------------------------

#: Bumped whenever a change alters the output for unchanged input.
#:
#: Written onto every candidate and every job. Without it, a candidate found in
#: the moderation queue is unreproducible: there is no way to tell which version
#: of the scoring produced a 72, and therefore no way to review the decision or
#: to re-run it. Treat this the way a schema version is treated.
ALGORITHM_VERSION = "1.0.0"


# -----------------------------------------------------------------------------
# Ingested data
# -----------------------------------------------------------------------------


class IncidentReport(BaseModel):
    """An approved report, as read from Firestore.

    A deliberate subset of the mobile app's ``IncidentReport``. The description
    text, the photograph URLs and the moderation notes are **not** read: this
    service clusters positions and categories, and pulling free text and images
    it has no use for into an analytics pipeline would be collecting more
    personal data than the job needs.

    ``reporter_id`` is read, but only to count *distinct reporters* per cluster —
    see ``risk_score``. It never leaves the service.
    """

    model_config = ConfigDict(frozen=True)

    id: str
    reporter_id: str
    type: IncidentType
    severity: IncidentSeverity
    latitude: float
    longitude: float
    status: ReportStatus
    #: When the incident happened, when the reporter said. Falls back to
    #: ``created_at`` in the pipeline, never silently to "now".
    occurred_at: datetime | None = None
    created_at: datetime | None = None

    @field_validator("latitude")
    @classmethod
    def _valid_latitude(cls, value: float) -> float:
        if not -90.0 <= value <= 90.0:
            raise ValueError(f"latitude out of range: {value}")
        return value

    @field_validator("longitude")
    @classmethod
    def _valid_longitude(cls, value: float) -> float:
        if not -180.0 <= value <= 180.0:
            raise ValueError(f"longitude out of range: {value}")
        return value

    @property
    def effective_time(self) -> datetime | None:
        """When the incident happened, as best as it is known."""
        return self.occurred_at or self.created_at


# -----------------------------------------------------------------------------
# Pipeline output
# -----------------------------------------------------------------------------


class Cluster(BaseModel):
    """A group of nearby reports, produced by DBSCAN."""

    model_config = ConfigDict(frozen=True)

    id: str
    centroid_latitude: float
    centroid_longitude: float
    #: Radius covering every member, floored at the minimum sane black spot size.
    radius_m: float
    report_ids: tuple[str, ...]
    #: How many *distinct* people reported here. See risk_score for why.
    distinct_reporters: int
    incident_types: tuple[IncidentType, ...]
    severities: tuple[IncidentSeverity, ...]

    @property
    def report_count(self) -> int:
        return len(self.report_ids)


class FrequentItemset(BaseModel):
    """One frequent itemset, as produced by ECLAT."""

    model_config = ConfigDict(frozen=True)

    items: tuple[str, ...]
    #: Number of transactions containing every item.
    support_count: int
    #: ``support_count / transaction_count``, in [0, 1].
    support: float


class RiskScore(BaseModel):
    """A cluster's score, with the components that produced it.

    The breakdown is stored, not just the total. A moderator being asked to
    publish a hazard on an algorithm's say-so is entitled to see why it scored
    what it did, and "72" on its own is not a reason.
    """

    model_config = ConfigDict(frozen=True)

    score: int = Field(ge=0, le=100)
    risk_level: RiskLevel
    components: dict[str, float]


class BlackSpotCandidate(BaseModel):
    """A proposed black spot. **Never** a published one.

    Written to ``blackSpotCandidates``, which the mobile app cannot read and no
    client can write. Publishing means an administrator creating a document in
    ``blackSpots`` — a separate, deliberate, audited act. That separation is a
    standing project rule: an algorithm must not be able to put a warning in
    front of users on its own.
    """

    model_config = ConfigDict(frozen=True)

    id: str
    latitude: float
    longitude: float
    geohash: str
    radius_m: float
    risk_level: RiskLevel
    severity_score: int = Field(ge=0, le=100)
    score_components: dict[str, float]
    category: BlackSpotCategory
    #: Always ``algorithm``. A candidate has exactly one provenance.
    source: BlackSpotSource = BlackSpotSource.ALGORITHM
    report_ids: tuple[str, ...]
    report_count: int
    distinct_reporters: int
    #: Human-readable patterns ECLAT found in this cluster.
    patterns: tuple[str, ...]
    algorithm_version: str = ALGORITHM_VERSION
    job_id: str


class AnalysisJob(BaseModel):
    """Metadata for one pipeline run.

    Recorded so a run is auditable and repeatable: which inputs, which
    parameters, which algorithm version, and what came out.
    """

    model_config = ConfigDict(frozen=True)

    id: str
    started_at: datetime
    finished_at: datetime | None = None
    algorithm_version: str = ALGORITHM_VERSION
    reports_ingested: int = 0
    reports_after_cleaning: int = 0
    duplicates_removed: int = 0
    clusters_found: int = 0
    candidates_written: int = 0
    #: Every tunable that affected the result, so a run can be reproduced.
    parameters: dict[str, float | int | str] = Field(default_factory=dict)
    status: str = "running"
    error: str | None = None
